Manually triggered GitHub Action for static checks using clang-tidy

williamfgc opened the pull request:
## Proposed changes

Adds a `static` manual workflow on GitHub Actions, triggered from the Actions tab. 
Get results from compiling with clang-tidy (takes 2 hours 15 min) on-demand on the develop branch only.
Update docker image to add clang-tidy dependency.
Related to #3941 as we address clang-tidy recommendations.
Needs merging so an admin can trigger the workflow.

## What type(s) of changes does this code introduce?

- Testing changes (e.g. new unit/integration/performance tests) static checks

### Does this introduce a breaking change?

- Yes

## What systems has this change been tested on?
Runs on GitHub Actions on my fork, must run on QMCPACK/qmcpack

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- No. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

williamfgc commented:
For reference this is the resulting workflow logs on my [fork](https://github.com/williamfgc/qmcpack/actions/runs/2339592357).

correaa commented:
The PR looks good to me. 
Depending on the number of compilation units, `"clang-tidy;-checks=*"` might produce very large logs to the point that github could truncate them.

There will be a flood of warnings at the beginning so we might want to define a protocol on how to deal with them, for example if it is decided that nothing will be done about a certain warning (for example, disable the warning globally vs. disabling check in that line `  // NOLINT` vs disabling specific check `  // NOLINT(specific-warning)` vs adding a useful comment ` // NOLINT(specific-warning) comment: implicit conversion is necessary here because bla...`.

I think we can safely disable this three warnings to start with because they don't make sense.

```
  -llvmlibc-callee-namespace,
  -llvmlibc-implementation-in-namespace,
  -llvmlibc-restrict-system-libc-headers,
```

Finally, as exceptions mount, it is better to have `.clang-tidy` configuration file that can sit in the root directory. For example: https://gitlab.com/correaa/boost-multi/-/blob/master/.clang-tidy

correaa reviewed (approved):
Look good to me. See comments in the PR thread about possible future issues.

ye-luo commented:
Need some documentation. I mean if I see a CI failure, how to reproduce clang-tidy manually.

correaa commented:
> Need some documentation. I mean if I see a CI failure, how to reproduce clang-tidy manually.



> Need some documentation. I mean if I see a CI failure, how to reproduce clang-tidy manually.

yes, even for reproducibility sake, as soon as the command line becomes more complicated that `-DCMAKE_CXX_CLANG_TIDY="clang-tidy;-checks=*"`, we will start needing a configuration file `.clang-tidy` that is part of the repository.

markdewing commented:
There already is a .clang-tidy file in the src/ directory
https://github.com/QMCPACK/qmcpack/blob/develop/src/.clang-tidy

correaa commented:
> There already is a .clang-tidy file in the src/ directory https://github.com/QMCPACK/qmcpack/blob/develop/src/.clang-tidy

ok, in that case the command in the CI can be simply `-DCMAKE_CXX_CLANG_TIDY="clang-tidy"`.

prckent reviewed (changes requested):
We should be using a current version of Clang right from the beginning. e.g. 13, since 14 is a bit too recent to be available everywhere. LLVM10 will be out of our 2 year support window in 4 months.

Given that we have so many warnings at the moment, have you found a convenient way to review them in the GitHub interface? My experience with the current CI logs is not good for any lengthy file.

williamfgc commented:
Changes:
- Added docs @ye-luo 
- Added Ubuntu22 image with Clang14 (default) for a recent clang-tidy @prckent 
- Added `.clang-tidy` with `readability-*` and `bugprone-*` as seen in #3941. @correaa

ye-luo commented on docs/github_actions.rst:
```diff
@@ -945,3 +945,30 @@ Generate code coverage reports once all tests have reported.
+
+**Note:** the current `.clang-tidy` configuration file is compatible with clang v14 and runs on the `williamfgc/qmcpack-ci:ubuntu22-openmpi` docker image. To run locally on a Linux system use: `docker run -it user williamfgc/qmcpack-ci:ubuntu22-openmpi /bin/bash` or refer to the :ref:`running_docker`` section.
+
+To build locally enaling `clang-tidy`` static checks defined in `qmcpack/.clang-tidy` use the CMake `-DCMAKE_CXX_CLANG_TIDY` option as follows:
```
`src/.clang-tidy` instead of `qmcpack/.clang-tidy`

williamfgc replied:
@ye-luo good point. @PDoakORNL @quantumsteve I'm unifying the `.clang-tidy` configuration. Are you OK removing the one under `/src`? The one at the top level would have similar checks.

ye-luo replied:
We should have only one configure file. I don't get why it should be under the root directory. If I understand correctly, clang-tidy handles C++ source code and all C++ code managed by QMCPACK is under src. Codes under external should not be subject to clang-tidy checks by QMCPACK.

ye-luo commented on docs/github_actions.rst:
```diff
@@ -945,3 +945,30 @@ Generate code coverage reports once all tests have reported.
+
+**Note:** the current `.clang-tidy` configuration file is compatible with clang v14 and runs on the `williamfgc/qmcpack-ci:ubuntu22-openmpi` docker image. To run locally on a Linux system use: `docker run -it user williamfgc/qmcpack-ci:ubuntu22-openmpi /bin/bash` or refer to the :ref:`running_docker`` section.
+
+To build locally enaling `clang-tidy`` static checks defined in `qmcpack/src/clang-tidy` use the CMake `-DCMAKE_CXX_CLANG_TIDY` option as follows:
```
`qmcpack/src/.clang-tidy` missing a dot.
"enaling" typo?

PDoakORNL commented:
Our .clang-tidy should be in src it should not be a reachable parent for the external_codes.  Additionally be aware that subdirectories can have there own .clang-tidy which may optionally inherit from its parent or not.  For instance in src/type_traits if am experimenting with the following 
```yaml
InheritParentConfig: true
CheckOptions: [ { key: readability-identifier-naming.StructIgnoredRegexp, value: '.*(_t|_impl)$'}
                ]
```
since its rather helpful to reading template metaprogramming to have types with _t and _impl and no have a ton of //NOLINT comments.

Some of the lost cause directories could also be squelched from the clang-tidy reporting by adding `InheritParentConfig: false` .clang-tidy in their directory.

correaa commented:
> since its rather helpful to reading template metaprogramming to have types with _t and _impl and no have a ton of //NOLINT comments.

I use that idiom in my naming and yet I didn't have to put anything special to the clangtidy.


> Some of the lost cause directories could also be squelched from the clang-tidy reporting by adding `InheritParentConfig: false` in .clang-tidy local to those directories.  

I didn't know about this option, seems useful (but not for external code) but for directories, (like tests).

PDoakORNL commented:
@correaa in our case we've defined type naming rules that disallow that for types.  They're specified as JustMixedCase.

williamfgc commented:
Thank you all for your input. I think I addressed the review comments. Let me know if there anything pending. We can all refer to the manually generated report after merge to start addressing issues, in particular those of `bugprone-*` type.

ye-luo commented on docs/github_actions.rst:
```diff
@@ -945,3 +945,30 @@ Generate code coverage reports once all tests have reported.
+
+====== ======== ======== 
+Static Analysis Workflow
+====== ======== ========
```
I don't see breaking lines based on spaces.

ye-luo commented on docs/github_actions.rst:
```diff
@@ -945,3 +945,30 @@ Generate code coverage reports once all tests have reported.
+
+**Note:** the current `.clang-tidy` configuration file is compatible with clang v14 and runs on the `williamfgc/qmcpack-ci:ubuntu22-openmpi` docker image. To run locally on a Linux system use: `docker run -it user williamfgc/qmcpack-ci:ubuntu22-openmpi /bin/bash` or refer to the :ref:`running_docker`` section.
+
+To build locally enaling `clang-tidy`` static checks defined in `qmcpack/src/.clang-tidy` use the CMake `-DCMAKE_CXX_CLANG_TIDY` option as follows:
```
`enaling` is still there.

ye-luo commented on docs/github_actions.rst:
```diff
@@ -945,3 +945,30 @@ Generate code coverage reports once all tests have reported.
+Static Analysis Workflow
+====== ======== ========
+
+A manually triggered workflow on the GitHub Actions tab can generate the required checks using the [clang-tidy](https://clang.llvm.org/extra/clang-tidy/) static analyzer. The current approach is to set checks in the `qmcpack/.clang-tidy` configuration file and run using `clang-tidy` v14 on GitHub Actions runners. The workflow is not part of CI, and it's currently used for reporting potential warn ...
```
`qmcpack/.clang-tidy` still

ye-luo commented:
Test this please

prckent commented:
Approved assuming that I didn't break the rst.

prckent commented:
Test this please

correaa commented:
ok, i see. i didn’t know about JustMixedCase. what is the difference with
CamelCase?

On Mon, May 23, 2022 at 9:27 AM Peter Doak ***@***.***> wrote:

> @correaa <https://github.com/correaa> in our case we've defined type
> naming rules that disallow that for types. They're specified as
> JustMixedCase.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/QMCPACK/qmcpack/pull/4021#issuecomment-1134888063>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXICU24JKPMEKXI3VNHN4LVLOWYTANCNFSM5WF6PQTQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
